www/_blog/2022-08-11-authentication-tutorial.mdx

```diff
@@ -0,0 +1,597 @@
+```
+
+Notice the action and method that correspond to the route we just added. Notice also the enctype being multipart, which matches what we are parsing in the handler. The above has a few attributes to do some client-side validation, but [in the full demo it is also handled on the server](https://github.com/kaleidawave/axum-shuttle-postgres-authentication-demo/blob/ba71a914055f312636581f5e82172b1078e7b9eb/src/authentication.rs#L124-L133). 
+
+We create a handler for this markup in the same way as done for our index with:
+
+```rust
+async fn get_signup(
+    Extension(templates): Extension<Templates>,
+) -> impl IntoResponse {
+    Html(templates.render("signup", &Context::new()).unwrap())
+}
+```
+
+We can add `signup` to the tera instance and then add both the get and post handlers to the router by adding it to the chain:
+
+```rust
+.route("/signup", get(get_signup).post(post_signup))
+```
+
+### Sessions
+
+Once signed up we want to save the logged-in state. We don't want the user to have to send their username and password for every request they make.
+
+### Cookies and session tokens
+
+Cookies help store state between browser requests. When a response is sent down with [Set-Cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie), then any subsequent requests the browser/client makes will send cookie information. We can then pull this information off of headers on requests on the server.
+
+Again these need to be safe. We don't want collisions/duplicates. We want it to be hard to guess. For these reasons, we will represent it as a 128-bit unsigned integer. This has 2^128 options, so a very low chance of a collision.
+
+We want to generate a "session token". We want the tokens to be cryptographically secure. Given a session id, we don't want to be able to guess the next one. aka if we have a global u128 and I have session 40129 then I can get 40130. We will use the [ChaCha](https://crates.io/crates/rand_chacha) algorithm/crate (we will add `cargo add rand_core rand_chacha`). [We can see that it does implement the crypto marker trait confirming it is valid for cryptographic scenarios](https://docs.rs/rand_chacha/0.3.1/rand_chacha/struct.ChaCha8Rng.html#impl-CryptoRng).
+
+This is unlike [Pseudo-random number generators where you can predict the next *random* number given a start point and the algorithm](https://www.youtube.com/watch?v=-h_rj2-HP2E). This could be a problem if we have our token we can get the session token of the person who logged in after us really easy and thus impersonate them.
+
+To start the random we have to set up a seed. We will be doing something similar to the creation of the Tera instance. We must wrap it in an arc and a mutex because generating new identifiers requires mutable access. We then start this in our main function with `let random = ChaCha8Rng::seed_from_u64(OsRng.next_u64());` and add it as an extension to our router chain with `.layer(Extension(Arc::new(Mutex::new(random))))`.

```

Comment:
Maybe we should rather have the code context from line ~188 again over the textual description?

Reply:
A second code snippet might be needed here to show the update for the `post_signup` function signature. (This relates to the `random` comment)